components/concurrency_manager/src/lib.rs

```diff
@@ -23,38 +23,115 @@ use std::{
     },
 };
 
+use lazy_static::lazy_static;
+use prometheus::{register_int_gauge, IntGauge};
+use thiserror::Error;
 use txn_types::{Key, Lock, TimeStamp};
 
 pub use self::{
     key_handle::{KeyHandle, KeyHandleGuard},
     lock_table::LockTable,
 };
 
+lazy_static! {
+    static ref MAX_TS_LIMIT_GAUGE: IntGauge = register_int_gauge!(
+        "tikv_concurrency_manager_max_ts_limit",
+        "Current value of max_ts_limit"
+    )
+    .unwrap();
+    static ref MAX_TS_GAUGE: IntGauge =
+        register_int_gauge!("tikv_concurrency_manager_max_ts", "Current value of max_ts").unwrap();
+}
+
 // Pay attention that the async functions of ConcurrencyManager should not hold
 // the mutex.
 #[derive(Clone)]
 pub struct ConcurrencyManager {
     max_ts: Arc<AtomicU64>,
+    max_ts_limit: Arc<AtomicU64>,
     lock_table: LockTable,
+
+    pub panic_on_invalid_max_ts: bool,
 }
 
 impl ConcurrencyManager {
     pub fn new(latest_ts: TimeStamp) -> Self {
+        Self::new_with_config(latest_ts, true)
+    }
+
+    pub fn new_with_config(latest_ts: TimeStamp, panic_on_invalid_max_ts: bool) -> Self {
         ConcurrencyManager {
             max_ts: Arc::new(AtomicU64::new(latest_ts.into_inner())),
+            max_ts_limit: Arc::new(AtomicU64::new(0)),
             lock_table: LockTable::default(),
+            panic_on_invalid_max_ts,
         }
     }
 
     pub fn max_ts(&self) -> TimeStamp {
         TimeStamp::new(self.max_ts.load(Ordering::SeqCst))
     }
 
+    /// Updates max_ts with a timestamp from PD TSO, bypassing the limit check.
+    /// TS from PD must be a valid, thus it doesn't need to be checked.
+    /// It has no effect if max_ts >= new_ts or new_ts is TimeStamp::max().
+    ///
+    /// # Safety
+    /// This method should only be used with timestamps obtained from PD TSO.
```

Comment:
```suggestion
    /// This method should only be used with timestamps obtained from PD TSO.
    /// WARN: An invalid timestamp would break the linearizability of the whole cluster.
```